[Woo POS] Fix for app crashing in payment flow on config change

samiuelson opened the pull request:
##  Fix for app crashing in payment flow on config change

<!-- Remember about a good descriptive title. -->

Closes: #12308 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

The app was crashing in the payment flow on config change (e.g. when changing the device's dark/light theme). Debugging revealed multiple problems:
1. The nav graph was not properly inflated and set to `NavController` in `WooPosCardReaderActivity` when the Activity was recreated. It was set only during view model initialization and was thus ignored in case Activity died and was recreated, but VM survived.
2. `WooPosCardReaderFacade` was setting `WooPosCardReaderFacade.paymentContinuation` to null when the Activity was dying. After activity recreation, the `paymentContinuation` prop was null, and the payment result was not delivered down to the `WooPosTotalsViewModel`. As a result, the "payment successful" screen was not shown, leaving the user in the Totals screen – quite misleading.

The above issues were fixed by:
1. Refactoring the event-style state rendering of the `WooPosCardReaderActivity` state to a static state exposed by the VM's property.
2. Not allowing the `paymentContinuation` to be nulled during Activity recreation.

Additionally, the `registerForActivityResult` was improved, making it offensive when passing payment result back to VM via `paymentContinuation`.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Start payment flow
4. Init config change, e.g. change from dark to light mode
5. Notice crash:
```kotlin
java.lang.IllegalArgumentException: No destination with ID 2131364278 is on the NavController's back stack. The current destination is null
```
### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->
I've tested the POS payment collection flow with and without triggering configuration change (device dark/light mode switch) and confirmed the flow completes as expected. Samsung Galaxy Tab, Android 14.

Note: The IPP flow itself remains untouched so it's not necessary to test against regression in the store management mode. It would be great to test the card reader connection flow and payment collection flows in POS mode extensively though.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/90072f11-a376-41e0-8c2e-62955673c1e6

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Fix nav graph setting when WooPosCardReaderActivity is recreated

Previously the nav graph was inflated and set only on the event received by the VM. As a result, the graph was null after activity recreation.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson committed:
Don't let payment continuation die during activity recreation

Previously `WooPosCardReaderFacade.paymentContinuation` was set to null inside `WooPosCardReaderFacade::onDestroy`. As a result, when payment result was received by `registerForActivityResult`'s callback it was not passed down to the `WooPosTotalsViewModel`.

samiuelson committed:
Make `registerForActivityResult`'s callback more offensive

We should not allow the state in which payment result is received, but `paymentContinuation` is null.

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Fix typo

samiuelson committed:
Remove ScopedViewModel supertype

samiuelson committed:
Convert prop's initializer to getter

samiuelson committed:
Merge branch 'trunk' into 12308-fix-crash-in-payment-flow-on-config-change

backwardstruck committed:
Merge branch 'trunk' into 12308-fix-crash-in-payment-flow-on-config-change

backwardstruck commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/cardreader/WooPosCardReaderActivity.kt:
```diff
@@ -26,7 +26,7 @@ class WooPosCardReaderActivity : AppCompatActivity(R.layout.activity_woo_pos_car
         ) as NavHostFragment
 
-        observeEvents(navHostFragment)
+        setupNavGraph(navHostFragment)
```
OK. I had suspected something like this was going on but I didn't come up with this solution in my [attempt](https://github.com/woocommerce/woocommerce-android/pull/12325).

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/cardreader/WooPosCardReaderFacade.kt:
```diff
@@ -38,13 +38,12 @@ class WooPosCardReaderFacade @Inject constructor(cardReaderManager: CardReaderMa
             }
 
-            paymentContinuation?.resume(paymentResult!!)
+            paymentContinuation!!.resume(paymentResult!!)
```
Just thinking about this change

I checked the code of that `registerForActivityResult` and looks like that they can return the result only of the starting activity is in between start and start, but I am wondering if there is a chance that the starting activity can be destroyed already after the callback called? Seems to be no as this will be run on the main thread, so looks a valid change to me

kidinov commented:
@samiuelson 

> The nav graph was not properly inflated and set to NavController in WooPosCardReaderActivity when the Activity was recreated. It was set only during view model initialization and was thus ignored in case Activity died and was recreated, but VM survived.

But it uses live data, so is the latest event supposed to be delivered after active recreation? 🤔 Was it delivered to late or what am I missing?

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/cardreader/WooPosCardReaderViewModel.kt:
```diff
@@ -1,33 +1,37 @@
-            }
+class WooPosCardReaderViewModel @Inject constructor(val savedStateHandle: SavedStateHandle) :
+    ViewModel() {
+    val cardReaderMode: WooPosCardReaderMode
```
I am wondering; maybe in order to bring some sense to the existence of this model, we should cache the value returned from savedStateHandle? Otherwise everytime `cardReaderMode` is hit we go to the bundle to take the value

samiuelson replied:
The `WooPosCardReaderActivity` serves both for reader connection and payment collection actions. My initial train of thought was that it's safer to avoid caching in case the Activity is started/destroyed continuously with different `WOO_POS_CARD_READER_MODE_KEY` values (`Connection`/`Payment`), while the VM survives and provides stale value 🤔

kidinov replied:
> Activity is started/destroyed continuously with different WOO_POS_CARD_READER_MODE_KEY values (Connection/Payment)

I recently added a "single_top" flag (maybe not merged yet), so maybe, yeah, it is safer to go this way. We don;t need the VM though then =)

samiuelson replied:
Keeping `orderId` validation, etc. out of Activity can be useful, but I agree we don't need to employ the VM for it.

kidinov replied:
Yeah, but I'd keep it anyway, as there is still some kind of logic in it

samiuelson commented:
> @samiuelson
> 
> > The nav graph was not properly inflated and set to NavController in WooPosCardReaderActivity when the Activity was recreated. It was set only during view model initialization and was thus ignored in case Activity died and was recreated, but VM survived.
> 
> But it uses live data, so is the latest event supposed to be delivered after active recreation? 🤔 Was it delivered to late or what am I missing?

Good question, @kidinov. The `_events` property in `ScopedViewModel` is instantiated using custom implementation of `MutableLivedata`:
```kotlin
val _event: MutableLiveData<Event> = MultiLiveEvent()
```
where
```kotlin
/**
 * A lifecycle-aware observable that sends only new updates after subscription, used for events like
 * navigation and Snackbar messages.
 *
 *
 * This avoids a common problem with events: on configuration change (like rotation) an update
 * can be emitted if the observer is active. This LiveData only calls the observable if there's an
 * explicit call to setValue() or call().
 *
 *
 * This is a mutation of SingleLiveEvent, which allows multiple observers. Once an observer marks the event as handled,
 * no other observers are notified and no further updates will be sent, similar to SingleLiveEvent.
 */
open class MultiLiveEvent<T : Event> : MutableLiveData<T>() {
    private val pending = AtomicBoolean(false)

    @MainThread
    override fun observe(owner: LifecycleOwner, observer: Observer<in T>) {
        super.observe(owner) { t ->
            if (pending.get()) {
                t.isHandled = true
                observer.onChanged(t)
                pending.compareAndSet(t.isHandled, false)
            }
        }
    }
}
```
Also, the docs above mention the implementation prevents replaying old events after subscription. So that's the reason the events were not re-emitted after `WooPosCardReaderActivity` was recreated.

kidinov commented:
@samiuelson Thanks for the explanation

🤦  yep, I completely missed that it's not lifedata we use but the thing for events

kidinov reviewed (approved):
LGTM!

kidinov commented:
@samiuelson My understanding that all the test you ran have to be added in the PR description though:

```
The author listed all the tests they ran
```

p1724156878421179-slack-C025A8VV728

samiuelson commented:
> @samiuelson My understanding that all the test you ran have to be added in the PR description though:
> 
> ```
> The author listed all the tests they ran
> ```
> 
> p1724156878421179-slack-C025A8VV728

I briefly described the tests I did under "Testing information".
